aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+    [(datetime.datetime(2019, 1, 1, 0, 0), datetime.datetime(2019, 1, 3, 0, 0)), (datetime.datetime(2019, 1, 3, 0, 0), datetime.datetime(2019, 1, 4, 0, 0))]
+
+    """  # noqa: E501
+    dt_from = start
+    dt_to = start + intv
+    while dt_to < end:
+        yield (dt_from, dt_to)
+        dt_from += intv
+        dt_to += intv
+
+    dt_to = min(dt_to, end)
+    if dt_from < end:
+        yield (dt_from, end)
+
+
+def get_start_and_end_from_data(data) -> tuple[datetime | None, datetime | None]:
+    """
+    Get the start and end times from the cloud function data.
+    """
+    if data:
+        dates = {}
+        if data.get('attributes'):
+            dates = data.get('attributes', {})
+        elif data.get('start') or data.get('end'):
+            dates = data
+        elif data.get('message'):
+            try:
+                dates = dict(json.loads(data['message']))
+            except ValueError:
+                dates = {}
+
+        logger.info(f'data: {data}, dates: {dates}')
+
+        s_raw = dates.get('start')
+        e_raw = dates.get('end')
+
+        # this should except if the start/end is in an invalid format
+        start = datetime.fromisoformat(s_raw) if s_raw else None
+        end = datetime.fromisoformat(e_raw) if e_raw else None
+
+        return (start, end)
+
+    return (None, None)
+
+
+def process_default_start_and_end(
+    start, end, interval=DEFAULT_RANGE_INTERVAL
+) -> tuple[datetime, datetime]:
+    """
+    Process the start and end times.
+    """
+    if not end and not start:
+        # start of today
+        end = datetime.now()
+        start = end - interval
+    elif not start:
+        start = end - interval
+    elif not end:
+        end = start + interval
+
+    assert isinstance(start, datetime) and isinstance(end, datetime)
+    return start, end
+
+
+def get_date_intervals_for(
+    start, end, interval=DEFAULT_RANGE_INTERVAL
+) -> Iterator[tuple[datetime, datetime]]:
+    """Process start and end times (and get defaults)"""
+    s, e = process_default_start_and_end(start, end)
+    return date_range_iterator(s, e, intv=interval)
+
+
+def read_secret(project_id: str, secret_name: str) -> str | None:
+    """Reads the latest version of a GCP Secret Manager secret.
+
+    Returns None if the secret doesn't exist."""
+
+    secret_manager = secretmanager.SecretManagerServiceClient()
+    secret_path = secret_manager.secret_path(project_id, secret_name)
+
+    response = secret_manager.access_secret_version(
+        request={'name': f'{secret_path}/versions/latest'}
+    )
+    return response.payload.data.decode('UTF-8')
+
+
+def get_hail_entry(
+    key: str,
+    topic: str,
+    service_id: str,
+    description: str,
+    cost: float,
+    currency_conversion_rate: float,
+    usage: float,
+    batch_resource: str,
+    start_time: datetime,
+    end_time: datetime,
+    labels: dict[str, str] = None,
+) -> dict[str, Any]:
+    """
+    Get well formed entry dictionary from keys
```

Comment:
This doesn't really explain what this function does or what parameters are expected.